add: bin deletion

Mesteery opened the pull request:
It is now possible to delete his snippet by providing a token.
This token is provided by the controller.
The snippet can be removed via a request `DELETE /<id>`
with the header `Authorization: Token <ADMIN_TOKEN>`.
The front calls this route when the user clicks on the delete button
(if the correct token is provided via a query (`/<id>?token=<ADMIN_TOKEN>`)).

Some scripts (js) have been refactorized.

x0rld commented on bin/assets/scripts/loc.js:
```diff
@@ -43,3 +48,25 @@ function handleHash() {
+      })
+      .catch((err) => {
+        alert(`Une erreur est survenue (${err}), votre snippet n'a donc pas pu être supprimé. Regarder la console pour plus d'informations.`);
+        console.error("Nous vous recommendons d'ouvrir une issue sur https://github.com/readthedocs-fr/bin-server si le problème persite.", err);
```
tu t'es fail c'est persiste je pense pour le message dans le console.error

Mesteery committed:
refactor(scripts): little refactor

Julien00859 commented on bin/controller.py:
```diff
@@ -119,10 +123,12 @@ def post_new():
             except KeyError:
                 raise ValueError("Parent does not exist")
+        if token and len(token) > 22:
+            raise ValueError("Token must not exceed 22 chars")
```
Est-ce qu'il y a une limitation technique qui fait qu'on ne doive pas excéder 22 caractères ? Si c'est le cas elle devrait être documenté dans le commit

Mesteery replied:
Il n'y a pas de limitation technique, ca éviterai juste de se retrouver avec des tokens qui font je ne sais pas combien de caractères. 22 est le nombre de caractères générés par `token_urlsafe(16)`

Julien00859 replied:
Alors on peut peut-être ajouter dans le message d'erreur "as of 16 random bytes base64 encoded"

Julien00859 commented on bin/controller.py:
```diff
@@ -70,10 +72,11 @@ def post_new():
+    :param maxusage: (form) optional maximum download of the snippet before it is deleted
+    :param lifetime: (form) optional time (defined in seconds) the snippet is keep in the database before it is deleted
+    :param parentid: (form) optional snippet id this new snippet is a duplicate of
+    :param token: (form) optional the "admin" token allows you to delete your bin
```
TIL optional is one n, thank you :)

Julien00859 commented on bin/controller.py:
```diff
@@ -146,15 +154,43 @@ def get_html(snippetid, ext=None):
+        raise bt.HTTPError(404, "Snippet not found")
+
+    if snippet.token != auth[1]:
+        raise bt.HTTPError(401, "Unauthorized")
```
[`secrets.compare_digest`](https://docs.python.org/3/library/secrets.html#secrets.compare_digest)

Julien00859 commented on bin/controller.py:
```diff
@@ -146,15 +154,43 @@ def get_html(snippetid, ext=None):
+    """
+    Delete a snippet
+
+    :param Authorization: (header) required "Authorization: Token <ADMIN_TOKEN>" the "admin" token
```
:+1:

Julien00859 commented on bin/controller.py:
```diff
@@ -146,15 +154,43 @@ def get_html(snippetid, ext=None):
         ext=ext,
         snippetid=snippetid,
         parentid=snippet.parentid,
+        token=token if snippet.token and token == snippet.token else None,
```
Il y a un problème de sécurité ici, je peux forger des requêtes `GET /snippet?token=xxx`, si on me génère un template où `token` est vide c'est que ce n'était pas le bon mot de passe. Je peux ainsi (sans jamais log aucune erreur ni avoir de 401) tester une myriade de token jusqu'à trouver le bon.

Julien00859 replied:
Je propose de systématiquement utiliser le token que l'utilisateur envoie, par design ce sera le token de la fonction `get_new_form` ou bien `None`. Si le gars s'amuse à forger des requête il aura juste l'air pas malin d'avoir un token qui ne marche pas.

Julien00859 commented on bin/models.py:
```diff
@@ -20,21 +20,23 @@ class Snippet:
         self.code = code  #: snippet text
         self.views_left = views_left  #: how many time this snippet can be retrieved again
         self.parentid = parentid  #: the original snippet this one is a duplicate of or an empty string
+        self.token = token #: the admin token of the snippet
```
La docstring de l'attribut doit documenter le cas spécial "None"

Julien00859 commented on bin/models.py:
```diff
@@ -20,21 +20,23 @@ class Snippet:
         :param maxusage: how many times this snippet can be retrieve before self-deletion
         :param lifetime: how long the snippet is saved before self-deletion
         :param parentid: the original snippet id this new snippet is a duplicate of, empty string for original snippet
+        :param token: the "admin" token of the snippet
```
La docstring doit expliquer le cas spécial "None"

Julien00859 commented on bin/models.py:
```diff
@@ -65,11 +69,20 @@ def get_by_id(cls, ident):
+        """
+        Delete the snippet from the database
+        """
+        database.delete(self.id)
```
Je suis d'avis de vérifier le token directement dans la méthode delete, qu'on n'ait jamais besoin d'utiliser le token en dehors de cette classe. Il faudrait lever un exception lorsque le token est invalide (ValueError ? Il faut voir laquelle serait la mieux, je ne pense pas qu'une exception type AccessError existe :/) et alors récupérer cette exception dans le controller et la changer en HTTPError.

Mesteery replied:
Personellement, je suis pas du même avis, je pense que l'authentification devrait seulement être gérée par le controller et non pas par le model

Julien00859 replied:
D'accord, attendons un 3e avis pour trancher (si laissons ainsi si 3e avis ne se présente pas)

Julien00859 replied:
ok on laisse ainsi

Julien00859 commented on bin/assets/scripts/loc.js:
```diff
@@ -43,3 +45,25 @@ function handleHash() {
+        console.error("Nous vous recommendons d'ouvrir une issue sur https://github.com/readthedocs-fr/bin-server si le problème persiste.", err);
+      });
+  });
+}
```
C'est dommage d'avoir *besoin* de javascript pour cette fonctionnalité, est-ce que ce ne serait pas possible d'avoir un formulaire avec méthode `DELETE` plutôt ? Le fait qu'on passe par le header Authorization risque d'être problématique mais j'aimerais qu'on explore cette possibilité (je sais qu'on a des users avec noscript)

Mesteery replied:
les seules méthodes possibles avec un form sont `get` et `post` malheureusement

Hokkaydo replied:
Il suffit d'ajouter un field supplémentaire contenant la "vraie" méthode et d'en récupérer la valeur au niveau du traitement.
Bon nombre de frameworks web permettent ceci

Mesteery replied:
c'est-a-dire ?

Hokkaydo replied:
Dans le formulaire au niveau html

Julien00859 replied:
Pour référence <https://softwareengineering.stackexchange.com/questions/114156/why-are-there-are-no-put-and-delete-methods-on-html-forms> (TL;DR: le W3C n'a jamais pris la peine de bosser dessus, donc ça supporte que GET et POST, oui c'est de la merde) mais le problème reste entier, c'est une fonctionnalité intéressante que les utilisateurs noscript n'auront pas moyen d'utiliser. Est-ce que ce serait envisageable de documenter qqpart de lisible pour l'utilisateur la commande `cron` équivalente pour supprimer le bin ?

Julien00859 replied:
@Hokkaydo même en passant par un framework le problème reste entier (voir pire) : on parle des utilisateurs noscript ici, les mêmes qui ne pourront pas charger le-dit framework.

Hokkaydo replied:
Mmh je vois
Tu as une autre idée dans ce cas ?

Julien00859 commented on bin/assets/scripts/newform.js:
```diff
@@ -2,6 +2,15 @@ const form = document.forms[0];
+    token.value.select();
+    document.execCommand('copy');
+  });
+})
```
:+1:

Julien00859 commented on bin/views/highlight.html:
```diff
@@ -8,6 +8,14 @@
+                    <path fill-rule="evenodd" d="M6 2a2 2 0 00-2 2v12a2 2 0 002 2h8a2 2 0 002-2V7.414A2 2 0 0015.414 6L12 2.586A2 2 0 0010.586 2H6zm1 8a1 1 0 100 2h6a1 1 0 100-2H7z" clip-rule="evenodd" />
+                </svg>
+            </button>
+            % end
```
Cf mon commentaire noscript, est-ce que ce serait envisageable d'avoir un pure formulaire html pour cette feature ?

Julien00859 reviewed (changes requested):
Alors le va et vient du token de sécurité mérite une explication dans le message de commit, ce n'est clairement pas trivial. J'aime beaucoup comment tu as fait, on n'a pas besoin d'une quelconque persistance niveau client ainsi c'est top, on pourrait même imaginer utiliser le localstorage sur la page highlight pour persister la pair snippetid/deletetoken et ainsi permettre à l'auteur (le seul qui à l'origine a eu un deletetoken non nul) à réouvrir le snippet plus tard dans son navigateur pour le supprimer. Je pense proposer ça dans une autre issue si on merge cette solution :+1: 

Ah ! Aussi, petit commentaire niveau naming, je trouve qu'appeler la variable qui contient le token pour supprimer un bin devrait ne pas juste être `token`. Ici dans le contexte de cette PR c'est évident qu'il s'agit de ce token mais j'imagine qu'un contributeur externe aura un peu plus de mal que nous ;)  Dans ma propre branche (que j'ai jamais push pcq elle marchait pas) je l'avais appelé "deltok" capitalisé en fonction du langage mais ce nom pue la merde aussi après coup :upside_down_face:

AntoineJT commented on bin/assets/scripts/loc.js:
```diff
@@ -43,3 +48,25 @@ function handleHash() {
+      })
+      .catch((err) => {
+        alert(`Une erreur est survenue (${err.message}), votre snippet n'a donc pas pu être supprimé. Regardez la console pour plus d'informations.`);
+        console.error("Nous vous recommendons d'ouvrir une issue sur https://github.com/readthedocs-fr/bin-server si le problème persiste.", err);
```
Une petite typo s'est glissée ici
```suggestion
        console.error("Nous vous recommandons d'ouvrir une issue sur https://github.com/readthedocs-fr/bin-server si le problème persiste.", err);
```

Julien00859 commented on bin/controller.py:
```diff
@@ -6,6 +6,8 @@
 import cgi
 import re
+from secrets import token_urlsafe, compare_digest
+import json
```
json est pas utilisé, il faut bouger secrets en dessous de metrics (ordre alphabétique)

Julien00859 commented on bin/controller.py:
```diff
@@ -119,11 +124,17 @@ def post_new():
+        bt.response.content_type = 'text/plain'
+        return f"/{snippet.id}.{ext}"
+    else:
+    	bt.redirect(f'/{snippet.id}.{ext}')
```
Ya un changement d'API ici, en 1.0.0 la route /new revoie systématiquement une redirection (code 3xx) ici maintenant *parfois* on renvoie une 200 avec l'url dans le body. Si on merge ça on doit bump MAJOR. J'y suis pas tellement favorable.

Julien00859 replied:
`?noredirect=1` à la place de ACCEPT

Julien00859 commented on bin/controller.py:
```diff
@@ -135,6 +146,8 @@ def get_html(snippetid, ext=None):
     :param ext: (path) optional language file extension, used to determine the highlight backend
 
+    :param token: (query) optional the "admin" token
+
```
Ya un retour à la ligne en trop

Julien00859 commented on bin/controller.py:
```diff
@@ -146,15 +159,43 @@ def get_html(snippetid, ext=None):
+    :raises HTTPError: code 401 when the token is incorrect
+    """
+    auth = bt.request.get_header('Authorization', '').split(None, 1)
+    if len(auth) != 2 or auth[0].lower() != 'token':
```
`casefold()` à la place de `lower`

Julien00859 replied:
limite même on met rien

Julien00859 commented on bin/models.py:
```diff
@@ -20,21 +20,23 @@ class Snippet:
         self.code = code  #: snippet text
         self.views_left = views_left  #: how many time this snippet can be retrieved again
         self.parentid = parentid  #: the original snippet this one is a duplicate of or an empty string
+        self.token = token #: the admin token of the snippet, `None` if the snippet has no "admin" token
```
En reStructuredText le inline code c'est \`\`None\`\`

Julien00859 commented on bin/models.py:
```diff
@@ -20,21 +20,23 @@ class Snippet:
     """
 
-    def __init__(self, ident, code, views_left, parentid):
+    def __init__(self, ident, code, views_left, parentid, token=None):
```
Pas de `=None` ici, ok pour le create mais pour le `__init__` on DOIT le donner, quitte à passer `None` explicitement.

Julien00859 commented on bin/models.py:
```diff
@@ -65,11 +69,20 @@ def get_by_id(cls, ident):
         parentid = snippet[b'parentid'].decode('ascii')
+        token = None
+        if b'token' in snippet:
+            token = snippet[b'token'].decode('ascii')
```
```suggestion
        token = snippet.get(b"token", b"").decode() or None
```

Julien00859 commented on bin/views/highlight.html:
```diff
@@ -8,6 +8,12 @@
     <body>
 {{!codehl}}
         <div class="controls">
+            <button class="control" id=delete-button {{ f"data-token=\"{token}\"" if token else "" }} title="Supprimer le snippet">
```
Il y a la fonction `define()` dans bottle pour vérifier si une variable existe, je sais pas si ça rendrait mieux de l'utiliser ou non dans ce contexte ??

Mesteery replied:
token est toujours défini mais peut valoir `None`, donc `defined()` ne gère pas ce cas ?

Mesteery commented on bin/assets/scripts/newform.js:
```diff
@@ -2,6 +2,42 @@ const form = document.forms[0];
+      }
+    }, { once: true });
+    return;
+  }
```
we should find a better solution

Julien00859 replied:
Honnêtement, si on trouve rien et que tu arrives à le faire marcher, on merge ainsi, on est d'accord que c'est pas du beau code mais balek ya que nous qui le lirons.

NoeMastrorillo commented on bin/controller.py:
```diff
@@ -122,10 +126,12 @@ def post_new():
-    snippet = models.Snippet.create(code, maxusage, lifetime, parentid)
+    snippet = models.Snippet.create(code, maxusage, lifetime, parentid, token)
     logger.info("New %s snippet of %s chars: %s", lang, len(code), snippet.id)
     bt.redirect(f'/{snippet.id}.{ext}')
```
```suggestion
    bt.redirect(f'/{snippet.id}.{ext}' + (f'?token={token}' if token else ''))
```

Mesteery replied:
je pense que c'est juste toi qui a pas bien compris le fonctionnement 😄 
- tu copies le token
- une fois le bin crée, l'utilisateur doit lui-même ajouter le token manuellement à l'url, `/blaba.js?token=...`

mais en soit ça pourrait être une bonne idée de faire ce que tu as suggéré. (j'attend l'avis d'une personne supplémentaire)

NoeMastrorillo replied:
Ouais j'avais pas bien lu effectivement.
Après pourquoi pas mais je trouve ça assez tiré par les cheveux.

Mesteery replied:
oui je sais bien mais...
> Vu la complexité énorme de cette PR. Je vais enlever l'enregistrement du token dans le localStorage. On discutera de ça dans une issue.
Alors, l'utilisateur devrait spécifier manuellement le token.

plutôt dans la PR le token était également enregistré dans le localStorage et donc la suppression était un peu plus "automatique", mais vu tous les problèmes que ça engendre et leurs complexités (t'as qu'à regarder tous les commentaires plus haut) je préfère rendre cette PR comme ça et qu'on en discute dans une issue à part de ce qu'on devrait faire.

Mesteery replied:
Je me rappelle, on n'avait pas fait ça, car l'utilisateur risquerait de copier son token quand il partagerait son bin.

NoeMastrorillo reviewed (changes requested):
De mon côté ça ne marche pas, j'ai bien le token au moment de créer le bin mais lorsqu'il est créé je n'ai pas le bouton pour delete.
J'ai pas trop suivi la discussion mais je pense que tu as oublié de mettre le token dans la query au moment de redirect sur le snippet après la création.

Julien00859 commented on bin/controller.py:
```diff
@@ -72,11 +74,12 @@ def post_new():
+    :param maxusage: (form) optional maximum download of the snippet before it is deleted
+    :param lifetime: (form) optional time (defined in seconds) the snippet is keep in the database before it is deleted
+    :param parentid: (form) optional snippet id this new snippet is a duplicate of
+    :param token: (form) optional the "admin" token allows you to delete your snippet
```
Je vais être un chieur mais... (et tu peux m'envoyer bouler) on pourrait corriger les typoes dans un commit à part ? :x

Julien00859 commented on bin/controller.py:
```diff
@@ -157,8 +164,35 @@ def get_html(snippetid, ext=None):
+    if not (snippet.token and secrets.compare_digest(snippet.token, auth[1])):
+        raise bt.HTTPError(401, "Unauthorized")
+
+    snippet.delete()
```
On peut ajouter un peu de log ici ?
```suggestion
    snippet.delete()
    logger.info("Snippet %s deleted by user", snippetid)
```

Mesteery committed:
feat: bin deletion

It is now possible to delete his snippet by providing a token.
This token is provided by the controller.
The snippet can be removed via a http request : DELETE /<id>
with the header Authorization: Token <ADMIN_TOKEN>.
The front calls this route when the user clicks on the delete button
if the token is provided via the query (/<id>?token=<token>).

Closes https://github.com/readthedocs-fr/bin-server/issues/98.

Co-authored-by: Antoine James Tournepiche <26577763+AntoineJT@users.noreply.github.com>
Co-authored-by: Julien Castiaux <Julien.castiaux@gmail.com>

Mesteery committed:
docs: correct typo

Julien00859 commented on bin/assets/scripts/loc.js:
```diff
@@ -1,45 +1,92 @@
+  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 20 20" fill="currentColor">
+    <path fill-rule="evenodd" d="M6 2a2 2 0 00-2 2v12a2 2 0 002 2h8a2 2 0 002-2V7.414A2 2 0 0015.414 6L12 2.586A2 2 0 0010.586 2H6zm1 8a1 1 0 100 2h6a1 1 0 100-2H7z" clip-rule="evenodd" />
+  </svg>
+</button>` + controls.innerHTML;
```
Il faudrait déplacer ce block dans le template bottle

Mesteery replied:
je l'ai déplacé ici exprès. Pour pas à avoir un bouton qui ne marche pas pour les noscript. (et surtout que c'est 100% géré par du JS)

Julien00859 replied:
Ouais je comprend la démarche... sauf que du coup on a un bout de templating d'un côté et un bout de templating de l'autre. Que penses-tu de laisser le bouton dans le template bottle mais avec un attribut `hidden` retiré dynamiquement par JS ? Ainsi on garde le html dans bottle mais le bouton n'est quand même pas visible aux noscripts :)
